### ISSUE TITLE
How to set litellm config?

### ISSUE THREAD

**Issue by NextGenOP** [Aider-AI/aider#949]
### Issue

I have litellm config that i usually use, how to load them with aider?

### Version and model info

Aider v0.45.1

**Comment by paul-gauthier** [comment]
Thanks for trying aider and filing this issue.

I'm not sure what you mean by "load litellm config"?

This doc may be helpful:

https://aider.chat/docs/llms.html

**Comment by paul-gauthier** [comment]
I'm going to close this issue for now, but feel free to add a comment here and I will re-open or file a new issue any time.

**Comment by gregid** [comment]
I would like to reopen this question and expand it slightly. Aider is using litellm for its models and by this I suppose it is using litellm python SDK. There is another option for running litellm and that's `litellm Proxy Server`. It exposes various models on e.g. localhost:4000 (default) and the models can be configured via config.yaml.
So the question (or feature request) is for aider to allow adding litellm server as an option for configuring and choosing the model to be used.
https://docs.litellm.ai/docs/simple_proxy

**Comment by NextGenOP** [comment]
> adding separate keys for litellm proxy

Why not put the key on the config file?

**Comment by gregid** [comment]
Unless I completely misunderstand your proposal I don't see a placement of the key as being an issue here. If we are using openai config we still end up with confusing `openai/anthropic/claude-3-5-sonnet`.
Your suggestion will only hide this "abomination" if we work in a set once and forget kind of manner.

**Comment by NextGenOP** [comment]
Emm. Isn't this the litellm config?
We may need to pass litellm into the aider model selection. But the api can be placed into config. With this you can have lot of different provider under single api.
https://docs.litellm.ai/docs/proxy/configs

**Comment by NextGenOP** [comment]
> Emm. Isn't this the litellm config? We may need to pass litellm into the aider model selection. But the api can be placed into config. With this you can have lot of different provider under single api. https://docs.litellm.ai/docs/proxy/configs

Edit: 
>LITELLM_API

Or maybe I misunderstood your proposal, is that litellm master key?

**Comment by gregid** [comment]
> > LITELLM_API
> 
> Or maybe I misunderstood your proposal, is that litellm master key?

Exactly, I suggest adding `LITELLM_API_BASE` and `LITELLM_API_KEY` so that we can then call our models like:
`litellm/anthropic/claude-3-5-sonnet` instead of: `openai/anthropic/claude-3-5-sonnet` and leaving openai key for other purposes

**Comment by NextGenOP** [comment]
I see.

**Comment by luizribeiro** [comment]
I'd love to be able to use an external instance of LiteLLM over the OpenAI-compatible API as well. I have LiteLLM setup on my server so I route all my LLM requests through it.

I tried setting it up with the OpenAI integration and changing the base URL, but `aider` attempts to use the `OPENROUTER_API_KEY` environment variable if I try to use a model such as `openrouter/anthropic/claude-3.5-sonnet` through my own LiteLLM instance.

Perhaps we could have a separate provider for LiteLLM which reads `LITELLM_API_KEY` and, if that provider is setup, all model requests go to LiteLLM?

If folks are ok with this approach I'd be happy to put up a PR.

**Comment by NextGenOP** [comment]
> Perhaps we could have a separate provider for LiteLLM which reads `LITELLM_API_KEY` and, if that provider is setup, all model requests go to LiteLLM?

I guess this method is easier than loading litellm config.

**Comment by luizribeiro** [comment]
Ok, I tried building this as I mentioned above and I'm running into issues because the `litellm` Python SDK still tries to route to the right provider for the model name. This is similar to the issue described on https://github.com/run-llama/llama_index/issues/13814 - which they seem to have decided that the right approach was to use the OpenAI API to talk to the LiteLLM Proxy.

I think there's two approaches we can take here if we want to allow `aider` to talk to a LiteLLM Proxy:
1. Just force to use the OpenAI API directly when using the new LiteLLM Proxy provider (comes with the downside of pulling the `openai` dependency to `aider`)
2. Put up a PR to the `litellm` Python SDK to allow for forcing a given provider, which seems less than ideal since it's a bit of bespoke logic being introduced to the framework just for this...

**Comment by NextGenOP** [comment]
> * Just force to use the OpenAI API directly when using the new LiteLLM Proxy provider (comes with the downside of pulling the `openai` dependency to `aider`)

my first thought is aider is using litellm proxy and uses openai sdk to call the model api.

are you using this?
[litellm.api_base](https://docs.litellm.ai/docs/set_keys#litellmapi_base)

**Comment by NextGenOP** [comment]
Turns out [this is the key](https://docs.litellm.ai/docs/providers/litellm_proxy)

1. I run `litellm --config litellm.yml`
2. I add [Required Variables](https://docs.litellm.ai/docs/providers/litellm_proxy#required-variables) to my ENV
3. I run `aider  --model litellm_proxy/one-model-from-litellm.yml`

Is it working for you?  
@luizribeiro 
@gregid

**Comment by gregid** [comment]
> Turns out [this is the key](https://docs.litellm.ai/docs/providers/litellm_proxy)
> 
> 1. I run `litellm --config litellm.yml`
> 2. I add [Required Variables](https://docs.litellm.ai/docs/providers/litellm_proxy#required-variables) to my ENV
> 3. I run `aider  --model litellm_proxy/one-model-from-litellm.yml`
> 
> Is it working for you? [@luizribeiro](https://github.com/luizribeiro) [@gregid](https://github.com/gregid)

Just to confirm this is the right solution to the problem - no need to make any changes to aider... maybe just adding this to documentation
